app/com/arpnetworking/rollups/RollupGenerator.java

```diff
@@ -310,60 +348,72 @@ private void handleNoMoreMetricsMessage(final NoMoreMetrics message) {
     }
 
     private LastDataPointMessage buildLastDataPointResponse(
-            final String metricName,
+            final String sourceMetricName,
+            final String rollupMetricName,
             final RollupPeriod period,
             final ImmutableSet<String> tags,
             final MetricsQueryResponse response,
             @Nullable final Throwable failure) {
         final LastDataPointMessage.Builder builder = new LastDataPointMessage.Builder()
-                .setMetricName(metricName)
+                .setSourceMetricName(sourceMetricName)
+                .setRollupMetricName(rollupMetricName)
                 .setPeriod(period)
                 .setTags(tags);
         if (failure != null) {
-            builder.setFailure(failure);
-        } else {
-            final Optional<MetricsQueryResponse.QueryResult> queryResult =
-                    response.getQueries()
+            return builder.setFailure(failure).build();
+        }
+        final Map<String, MetricsQueryResponse.QueryResult> queryResults =
+                response.getQueries()
                         .stream()
                         .flatMap(query -> query.getResults().stream())
-                        .findFirst();
-
-            // If there are no queries or query results, something has gone wrong, but an empty
-            // result itself does not indicate failure.
-            if (!queryResult.isPresent()) {
-                builder.setFailure(new Exception("Unexpected query results."));
-            } else {
-                final Instant lastDataPointTime =
-                        queryResult
-                        .flatMap(qr -> qr.getValues().stream().findFirst())
-                        .map(DataPoint::getTime)
-                        .orElse(null);
-                builder.setLastDataPointTime(lastDataPointTime);
-            }
+                        .collect(ImmutableMap.toImmutableMap(
+                                MetricsQueryResponse.QueryResult::getName,
+                                Function.identity()
+                        ));
+
+        // Query results should *only* contain the source and destination metric.
+        if (queryResults.size() != 2 || !queryResults.containsKey(sourceMetricName) || !queryResults.containsKey(rollupMetricName)) {
+            return builder.setFailure(new Exception("Unexpected query results.")).build();
         }
+
+        // Set source time, if any.
+        Optional.ofNullable(queryResults.get(sourceMetricName))
```

Comment:
The value should not be null after the check above. Right? If we're going to enforce the preconditions, we should simplify the resulting code.